File: python/238-product-of-array-except-self.py
from typing import List


class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        # ans, temp = [1], 1
        # for num in reversed(nums[1:]):
        #     temp *= num
        #     ans.insert(0, temp)
        # temp = 1
        # for i in range(len(nums) - 1):
        #     temp *= nums[i]
        #     ans[i + 1] *= temp
        # return ans

        new_array = []
        prefix = 1
        for num in nums:
            new_array.append(prefix)
            prefix = prefix * num
        postfix = 1
        for i in range(len(nums) - 1, -1, -1):
            new_array[i] = new_array[i] * postfix
            postfix = postfix * nums[i]
        return new_array


print(Solution().productExceptSelf([1, 2, 3, 4]))

File: medium/238-product-of-array-except-self.js
const productExceptSelf = (nums) => {
  const ans = new Array(nums.length).fill(1);
  console.log(ans);
  let left = 1;
  for (let i = 0; i < nums.length; i++) {
    ans[i] *= left;
    left *= nums[i];
  }
  console.log(ans);
  let right = 1;
  for (let i = nums.length - 1; i >= 0; i--) {
    ans[i] *= right;
    right *= nums[i];
  }
  return ans;
};

console.log(productExceptSelf([1,2,3,4]));